price/src/price_tracker.rs

```diff
@@ -0,0 +1,135 @@
+use anyhow::{anyhow, Error, Result};
+use chrono::{DateTime, Duration, Utc};
+use file_store::{FileInfo, FileStore, FileType, Settings};
+use futures::stream::{StreamExt, TryStreamExt};
+use helium_proto::{BlockchainTokenTypeV1, Message, PriceReportV1};
+use std::collections::HashMap;
+use tokio::sync::watch;
+
+type Prices = HashMap<BlockchainTokenTypeV1, u64>;
+
+#[derive(Clone)]
+pub struct PriceTracker {
+    receiver: watch::Receiver<Prices>,
+}
+
+impl PriceTracker {
+    pub async fn start(
+        settings: &Settings,
+        shutdown: triggered::Listener,
+    ) -> Result<(Self, impl std::future::Future<Output = Result<()>>)> {
+        let file_store = FileStore::from_settings(settings).await?;
+        let (sender, receiver) = watch::channel(Prices::new());
+        let initial_timestamp = get_initial_prices(&file_store, &sender).await?;
+
+        let handle =
+            tokio::spawn(async move { run(file_store, sender, initial_timestamp, shutdown).await });
+
+        Ok((PriceTracker { receiver }, async move {
+            match handle.await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(err)) => Err(err),
+                Err(err) => Err(Error::from(err)),
+            }
+        }))
+    }
+
+    pub async fn price(&self, token_type: &BlockchainTokenTypeV1) -> Result<u64> {
+        self.receiver
+            .borrow()
+            .get(token_type)
+            .cloned()
+            .ok_or_else(|| anyhow!("price not available"))
+    }
+}
+
+async fn run(
+    file_store: FileStore,
+    sender: watch::Sender<Prices>,
+    mut after: DateTime<Utc>,
+    shutdown: triggered::Listener,
+) -> Result<()> {
+    let mut trigger = tokio::time::interval(std::time::Duration::from_secs(30));
+
+    loop {
+        let shutdown = shutdown.clone();
+
+        tokio::select! {
+            _ = shutdown => {
+                tracing::info!("PriceTracker: shutting down");
+                break;
+            }
+            _ = trigger.tick() => {
+                let timestamp = process_files(&file_store, &sender, after).await?;
+                after = timestamp.unwrap_or(after);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn get_initial_prices(
```

Comment:
nit: the `get` in 'get_initial_prices' suggests that the function will be returning price information but instead a timestamp is returned.  I get that it is actually working out the initial prices but maybe renaming to calc_initial_prices is a better fit

Reply:
Good call, fixed 23b8f58